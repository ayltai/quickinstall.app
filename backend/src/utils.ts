import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
