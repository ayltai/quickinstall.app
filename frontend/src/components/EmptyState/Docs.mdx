import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './EmptyState.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use empty states

Empty states provides an opportunity to show explanation or guidance to help users progress.

When using empty states,
* Explain clearly the steps the users need to take to progress

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Examples'
    includePrimary={false} />
