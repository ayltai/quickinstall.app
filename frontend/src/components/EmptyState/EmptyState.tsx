import { ErrorOutline, } from '@mui/icons-material';
import Stack from '@mui/material/Stack';
import React, { type FC, } from 'react';

import type { EmptyStateProps, } from './EmptyState.types';

/**
 * A empty state indicates no items or data to show.
 * @param children The content of the component
 * @param rest Other props
 */
export const EmptyState : FC<EmptyStateProps> = ({
    children,
    ...rest
}) => (
    <Stack
        padding={2}
        display='flex'
        alignItems='center'
        textAlign='center'
        {...rest}>
        <ErrorOutline sx={{
            fontSize : 128,
        }} />
        {children}
    </Stack>
);
