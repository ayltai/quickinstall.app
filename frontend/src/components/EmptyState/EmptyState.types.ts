import type { HTMLAttributes, ReactNode, } from 'react';

export type EmptyStateProps = HTMLAttributes<HTMLDivElement> & {
    children?        : ReactNode,
    [ key : string ] : any,
};
