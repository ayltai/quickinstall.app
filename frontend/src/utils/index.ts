import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
