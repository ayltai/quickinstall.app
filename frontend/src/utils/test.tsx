import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
