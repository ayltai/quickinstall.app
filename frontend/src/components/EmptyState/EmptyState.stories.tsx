import { Typography, } from '@mui/material';
import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';

import { EmptyState, } from './EmptyState';

const meta : Meta<typeof EmptyState> = {
    component : EmptyState,
    title     : 'Components/Data Presentations/Empty State',
    argTypes  : {
        children : {
            description : 'The content of the component',
        },
    },
};

export default meta;

type Story = StoryObj<typeof EmptyState>;

export const Default : Story = {
    args : {
        children : <Typography>No results</Typography>,
    },
};
