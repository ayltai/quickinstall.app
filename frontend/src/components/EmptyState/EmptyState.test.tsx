import React from 'react';

import { render, } from '../../utils/test';
import { EmptyState, } from './EmptyState';

describe('<EmptyState />', () => {
    it('renders correctly', () => expect(render(<EmptyState>Content</EmptyState>)).toMatchSnapshot());
});
