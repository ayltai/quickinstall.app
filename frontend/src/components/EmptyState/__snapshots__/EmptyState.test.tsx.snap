// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EmptyState /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-ttb62z-MuiStack-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ykxrwf-MuiSvgIcon-root"
          data-testid="ErrorOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
          />
        </svg>
        Content
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-ttb62z-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ykxrwf-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
        />
      </svg>
      Content
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
